commands/system/help.js

```diff
@@ -16,67 +17,98 @@ class Help extends Command {
+      if (level < this.client.levelCache[cm.conf.permLevel]) return;
+      embed = new MessageEmbed(embedPreset)
+        .setTitle(cm.help.name.toProperCase())
+        .setDescription(`\u200B\t${cm.help.description}`)
+        .addField("Command details", cm.help.extended, false)
+        .addField("Command usage", `\`${cm.help.usage}\``, false)
+        .addField("Command aliases", cm.conf.aliases.length == 0 ? "None" : cm.conf.aliases.join(", "), false);
+      
+      return message.channel.send({ embed });
     } else {
-      let n = 0;
-      sorted.forEach(c => {
-        if (c.help.category.toLowerCase() === type.toLowerCase()) {
-          n++;
+      let currentCategory = "";
+      const sorted = this.client.commands.sort((p, c) => p.help.category > c.help.category ? 1 :  p.help.name > c.help.name && p.help.category === c.help.category ? 1 : -1 );
+      embed = new PaginationEmbed(embedPreset)
+        .setAuthorisedUser(message.author)
+        .setChannel(message.channel)
+        .setElementsPerPage(perpage)
+        .showPageIndicator(true)
+        .setTimeout(60 * 1000);
+      
+      const isPage = !isNaN(parseInt(type));
+      const isDefault = !type || isPage;
+
+      if (isDefault) {
+        const output = [];
+        const title = "Command category list";
+        const description = `Use \`${message.settings.prefix}help <category>\` to find commands for a specific category`;
+        const tip = `\nTo skip to a page for more categories:\n\tReact with ↗; or\n\tDo \`${message.settings.prefix}help <page-num>\``;
+        const categories = sorted.filter(c => !(level < 10 && c.help.category == "Owner") || !(c.help.category === "NSFW" && !message.channel.nsfw)).map(c => {
+          const cat = c.help.category.toProperCase();
+          if (currentCategory !== cat && isDefault) {
+            currentCategory = cat;
+            return `\`${message.settings.prefix}help ${cat.toLowerCase()}\` | Shows ${cat} commands`;
+          }
+        });
+        
+        for (let i = 0; i < categories.length; i++) {
+          if (categories[i] === undefined) continue;
+          output.push(categories[i]);
         }
-      });
-    
-      let output = "";
-      let num = 0;
-      const pg = parseInt(page) && parseInt(page) <= Math.ceil(n / perpage) ? parseInt(page) : 1;
-      for (const c of sorted.values()) {
-        if (c.help.category.toLowerCase() === type.toLowerCase()) {
-          if (c.help.category === "Owner" && level < 10 ) return;
-          if (c.help.category === "NSFW" && !message.channel.nsfw) return;
-          if (num < perpage * pg && num > perpage * pg - (perpage + 1)) {
+
+        embed
+          .setTitle(title)
+          .setDescription([description, output.length > perpage ? tip : ""])
+          .setArray(output)
+          .formatField("Categories", el => el);
+
+        if (isPage) embed.setPage(parseInt(type));
+      } else {
+        let n = 0;
+        sorted.forEach(c => {
```

Comment:
You could use this, which achieves the same goal but it's more performant. Check docs [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce).

```javascript
const lowerCaseCategory = type.toLowerCase();
const n = sorted.reduce((acc, cur) => c.help.category.toLowerCase() === lowerCaseCategory ? cur + 1 : cur, 0);
```